### PR TITLE
Decouple state handling from plugin options.

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1630762

Signed-off-by: Sergio Schvezov <sergio.schvezov@ubuntu.com>

### CODE REVIEW

**Comment by kyrofa** [comment]
(and an integration test)

File: snapcraft/internal/pluginhandler.py [hunk 1]
```diff
@@ -190,7 +190,7 @@ def is_dirty(self, step):
             # YAML (taken from self.code.options). If they've changed, then
             # this step is dirty and needs to run again.
             if state.properties != state.properties_of_interest(
-                    self.code.options):
+                    self.code.options.__dict__):
```
Comment (josepht) [inline snapcraft/internal/pluginhandler.py]: I think we should use `vars()` rather than `__dict__`

File: snapcraft/internal/pluginhandler.py [hunk 2]
```diff
@@ -273,7 +273,9 @@ def pull(self, force=False):
         self.notify_part_progress('Pulling')
         self.code.pull()
         self.mark_done('pull', states.PullState(
-            self.pull_properties, self.code.options, self._project_options))
+            self.pull_properties,
+            self.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/internal/pluginhandler.py]: ditto

File: snapcraft/internal/pluginhandler.py [hunk 3]
```diff
@@ -302,7 +304,9 @@ def build(self, force=False):
         self.notify_part_progress('Building')
         self.code.build()
         self.mark_done('build', states.BuildState(
-            self.build_properties, self.code.options, self._project_options))
+            self.build_properties,
+            self.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/internal/pluginhandler.py]: ditto

File: snapcraft/internal/pluginhandler.py [hunk 4]
```diff
@@ -346,7 +350,9 @@ def fixup_func(file_path):
         # dependencies here too
 
         self.mark_done('stage', states.StageState(
-            snap_files, snap_dirs, self.code.options, self._project_options))
+            snap_files, snap_dirs,
+            self.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/internal/pluginhandler.py]: ditto

File: snapcraft/internal/pluginhandler.py [hunk 5]
```diff
@@ -411,7 +417,8 @@ def prime(self, force=False):
         dependency_paths = (part_dependency_paths | staged_dependency_paths |
                             system_dependency_paths)
         self.mark_done('prime', states.PrimeState(
-            snap_files, snap_dirs, dependency_paths, self.code.options,
+            snap_files, snap_dirs, dependency_paths,
+            self.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/internal/pluginhandler.py]: ditto

File: snapcraft/tests/test_pluginhandler.py [hunk 6]
```diff
@@ -1419,7 +1419,7 @@ def test_prime_is_dirty(self):
         self.handler.code.options.snap = ['foo']
         self.handler.mark_done(
             'prime', states.PrimeState(
-                set(), set(), set(), self.handler.code.options))
+                set(), set(), set(), self.handler.code.options.__dict__))
```
Comment (josepht) [inline snapcraft/tests/test_pluginhandler.py]: ditto

File: snapcraft/tests/test_pluginhandler.py [hunk 7]
```diff
@@ -1443,7 +1443,7 @@ def test_stage_is_dirty(self):
         self.handler.code.options.stage = ['foo']
         self.handler.mark_done(
             'stage', states.StageState(
-                set(), set(), self.handler.code.options))
+                set(), set(), self.handler.code.options.__dict__))
```
Comment (josepht) [inline snapcraft/tests/test_pluginhandler.py]: ditto

File: snapcraft/tests/test_pluginhandler.py [hunk 8]
```diff
@@ -1468,7 +1468,7 @@ def test_build_is_dirty_from_options(self):
         self.handler.code.options.foo = ['bar']
         self.handler.mark_done(
             'build', states.BuildState(self.handler.build_properties,
-                                       self.handler.code.options,
+                                       self.handler.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/tests/test_pluginhandler.py]: ditto

File: snapcraft/tests/test_pluginhandler.py [hunk 9]
```diff
@@ -1486,7 +1486,7 @@ def test_build_is_dirty_from_options(self):
     def test_build_is_dirty_from_project(self, mock_enable_cross_compilation):
         self.handler.mark_done(
             'build', states.BuildState(self.handler.build_properties,
-                                       self.handler.code.options,
+                                       self.handler.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/tests/test_pluginhandler.py]: ditto ;)

File: snapcraft/tests/test_pluginhandler.py [hunk 10]
```diff
@@ -1515,7 +1515,7 @@ def test_pull_is_dirty_from_options(self):
         self.handler.code.options.foo = ['bar']
         self.handler.mark_done(
             'pull', states.PullState(self.handler.pull_properties,
-                                     self.handler.code.options,
+                                     self.handler.code.options.__dict__,
```
Comment (josepht) [inline snapcraft/tests/test_pluginhandler.py]: ditto :insane_smile:

File: snapcraft/tests/test_states_prime.py [hunk 11]
```diff
@@ -38,8 +38,8 @@ class Project:
         self.project = Project()
 
         self.state = snapcraft.internal.states.PrimeState(
-            self.files, self.directories, self.dependency_paths, self.options,
-            self.project)
+            self.files, self.directories, self.dependency_paths,
+            self.options.__dict__, self.project)
```
Comment (josepht) [inline snapcraft/tests/test_states_prime.py]: hey whadayaknow ditto

File: snapcraft/tests/test_states_prime.py [hunk 12]
```diff
@@ -50,22 +50,22 @@ def test_representation(self):
 
     def test_comparison(self):
         other = snapcraft.internal.states.PrimeState(
-            self.files, self.directories, self.dependency_paths, self.options,
-            self.project)
+            self.files, self.directories, self.dependency_paths,
+            self.options.__dict__, self.project)
```
Comment (josepht) [inline snapcraft/tests/test_states_prime.py]: ditto-re-no-diddle-oh

File: snapcraft/tests/test_states_prime.py [hunk 13]
```diff
@@ -50,22 +50,22 @@ def test_representation(self):
...
 
         self.assertTrue(self.state == other, 'Expected states to be identical')
 
     def test_comparison_not_equal(self):
         others = [
             snapcraft.internal.states.PrimeState(
                 set(), self.directories, self.dependency_paths,
-                self.options, self.project),
+                self.options.__dict__, self.project),
```
Comment (josepht) [inline snapcraft/tests/test_states_prime.py]: ditto

File: snapcraft/tests/test_states_prime.py [hunk 14]
```diff
@@ -50,22 +50,22 @@ def test_representation(self):
...
             snapcraft.internal.states.PrimeState(
                 self.files, set(), self.dependency_paths,
-                self.options, self.project),
+                self.options.__dict__, self.project),
```
Comment (josepht) [inline snapcraft/tests/test_states_prime.py]: ditto

File: snapcraft/tests/test_states_prime.py [hunk 15]
```diff
@@ -50,22 +50,22 @@ def test_representation(self):
...
             snapcraft.internal.states.PrimeState(
                 self.files, self.directories, set(),
-                self.options, self.project),
+                self.options.__dict__, self.project),
```
Comment (josepht) [inline snapcraft/tests/test_states_prime.py]: ditto

File: snapcraft/tests/test_states_stage.py [hunk 16]
```diff
@@ -37,7 +37,7 @@ class Project:
         self.project = Project()
 
         self.state = snapcraft.internal.states.StageState(
-            self.files, self.directories, self.options, self.project)
+            self.files, self.directories, self.options.__dict__, self.project)
```
Comment (josepht) [inline snapcraft/tests/test_states_stage.py]: ditto

File: snapcraft/tests/test_states_stage.py [hunk 17]
```diff
@@ -48,18 +48,18 @@ def test_representation(self):
 
     def test_comparison(self):
         other = snapcraft.internal.states.StageState(
-            self.files, self.directories, self.options, self.project)
+            self.files, self.directories, self.options.__dict__, self.project)
```
Comment (josepht) [inline snapcraft/tests/test_states_stage.py]: ditto

File: snapcraft/tests/test_states_stage.py [hunk 18]
```diff
@@ -48,18 +48,18 @@ def test_representation(self):
...
     def test_comparison_not_equal(self):
         others = [
-            snapcraft.internal.states.StageState(set(), self.directories,
-                                                 self.options, self.project),
-            snapcraft.internal.states.StageState(self.files, set(),
-                                                 self.options, self.project),
-            snapcraft.internal.states.StageState(self.files, self.directories,
-                                                 None, self.project),
+            snapcraft.internal.states.StageState(
+                set(), self.directories, self.options.__dict__, self.project),
```
Comment (josepht) [inline snapcraft/tests/test_states_stage.py]: ditto

File: snapcraft/tests/test_states_stage.py [hunk 19]
```diff
@@ -48,18 +48,18 @@ def test_representation(self):
...
+            snapcraft.internal.states.StageState(
+                self.files, set(), self.options.__dict__, self.project),
```
Comment (josepht) [inline snapcraft/tests/test_states_stage.py]: ditto

File: snapcraft/internal/pluginhandler.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/internal/pluginhandler.py]: Even though this code is transitional I will do it ;-)

I personally prefer __dict__

**Comment by sergiusens** [comment]
El 06/10/16 a las 11:32, Joe Talbott escribió:
>
> *@josepht* commented on this pull request.
>
> Should there be a test to ensue that |schema_properties| is still an 
> attribute of the instance so we don't break that in the future?
>

That is what test_representation does already ;-)

**Comment by josepht** [comment]
Sorry I must've missed it.

**Comment by sergiusens** [comment]
done @josepht

**Comment by sergiusens** [comment]
Not related error
```
======================================================================
FAIL: demos_tests.test_downloader_with_wiki_parts.DownloaderWithWikiPartsTestCase.test_downloader_with_wiki_parts
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/adt-run.7A3uZD/build.Ov2/real-tree/snaps_tests/demos_tests/test_downloader_with_wiki_parts.py", line 33, in test_downloader_with_wiki_parts
    '/snap/bin/downloader.test', expected, flags=re.DOTALL)
  File "/tmp/adt-run.7A3uZD/build.Ov2/real-tree/snaps_tests/__init__.py", line 239, in assert_command_in_snappy_testbed_with_regex
    self.assertThat(output, MatchesRegex(expected_regex, flags=flags))
  File "/usr/lib/python3/dist-packages/testtools/testcase.py", line 435, in assertThat
    raise mismatch_error
testtools.matchers._impl.MismatchError: '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">\n<html><head>\n<title>302 Found</title>\n</head><body>\n<h1>Found</h1>\n<p>The document has moved <a href="https://www.ubuntu.com/">here</a>.</p>\n<hr>\n<address>Apache/2.2.22 (Ubuntu) Server at www.ubuntu.com Port 80</address>\n</body></html>\n' does not match /.*<title>The leading OS for PC, tablet, phone and cloud | Ubuntu</title>.*/
```

Landing.

**Pull request merged**